lib/loglimit/loglimit.go

```
@@ -0,0 +1,198 @@
+	// by default.
+	Clock clockwork.Clock
+}
+
+// checkAndSetDefaults verifies configuration and sets defaults
+func (c *Config) checkAndSetDefaults() error {
+	if c.LogSubstrings == nil {
+		return trace.BadParameter("missing parameter LogSubstrings")
+	}
+	if c.ChannelSize < 0 {
+		return trace.BadParameter("ChannelSize must be at least 0")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	return nil
+}
+
+// LogLimiter deduplicates logs over a certain time window.
+type LogLimiter struct {
+	// config is the log limiter config.
+	config Config
+	// entryCh is used to send log entries to the log limiter.
+	entryCh chan *entryInfo
+	// windows is a mapping from log substring to an active
+	// time window.
+	windows map[string]*entryInfo
+	// cleanupCh is used in tests to trigger `cleanup`.
+	cleanupCh chan chan struct{}
+}
+
+// entryInfo contains information about a certain log entry.
+// This is information is used to deduplicate a log entry
+// reported within a certain time window.
+type entryInfo struct {
+	// entry is the logger entry.
+	entry *log.Entry
+	// level is the log level at which this log entry should be logged.
+	level log.Level
+	// message is the message message.
+	message string
+	// time is the at which the log reported.
+	time time.Time
+	// occurrences are the occurrences of logs entries (that share
+	// the same log substring) within a time window.
+	occurrences int
+}
+
+// New creates a new log limiter.
+func New(config Config) (*LogLimiter, error) {
+	if err := config.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &LogLimiter{
+		config:    config,
+		entryCh:   make(chan *entryInfo, config.ChannelSize),
+		windows:   make(map[string]*entryInfo, len(config.LogSubstrings)),
+		cleanupCh: make(chan chan struct{}),
+	}, nil
+}
+
+// Log sends a log to the log limiter.
+func (l *LogLimiter) Log(entry *log.Entry, level log.Level, args ...any) {
+	l.entryCh <- &entryInfo{
```

Adding to the comment above, should we make sure the limiter is running before blocking here?

---

Hmm, what should we do if the limiter is not running?

---

We could panic, swallow, log synchronously, error, or some combination of these. I'd say blocking indefinitely is worse than panicking, due to effects it may cause how difficult to diagnose it would be.

---

bfa26754eed112fc1bcc884cdae57c0e13727b05 removes the `entryCh` in favor of a mutex, which should address this issue.